framework/src/base/NonlinearSystem.C

```diff
@@ -1786,16 +1794,49 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
       {
         _fe_problem.reinitNodeFace(node, boundary_id, 0);
 
-        for (std::vector<NodalBC *>::const_iterator it = _bcs[0].getNodalBCs(boundary_id).begin(); it != _bcs[0].getNodalBCs(boundary_id).end(); ++it)
+        const std::vector<NodalBC *> & bcs = _bcs[0].getNodalBCs(boundary_id);
+        for (std::vector<NodalBC *>::const_iterator it = bcs.begin(); it != bcs.end(); ++it)
         {
           NodalBC * bc = *it;
-          if (bc->shouldApply() && bc->variable().isNodalDefined())
-            zero_rows.push_back(bc->variable().nodalDofIndex());
+
+          // Get the list of variables to which this BC is actually
+          // coupled.  We should not need to call
+          // computeOffDiagJacobian() for variables the BC is not
+          // coupled to.  Note: this list does not contain bc->variable.number().
+          const std::vector<MooseVariable *> & coupled_moose_vars = bc->getCoupledMooseVars();
+
+          // Extract set of coupled MooseVariable numbers, also
+          // sneakily insert bc->variable.number() to make the logic
+          // below a bit simpler.
+          std::set<unsigned int> involved_moose_var_nums;
+          for (unsigned int var=0; var < coupled_moose_vars.size(); ++var)
```

Comment:
We should be able to figure this out outside of the node loop... it may not take that much time... but on the other had `insert()`ing into a `set` definitely isn't free..